File: backend-springboot/src/main/java/br/com/portalproposta/backend/service/UsuarioService.java
package br.com.portalproposta.backend.service;

import br.com.portalproposta.backend.exception.AppBadRequestException;
import br.com.portalproposta.backend.exception.AppNotFoundException;
import br.com.portalproposta.backend.exception.AppUnauthorizedException;
import br.com.portalproposta.backend.infrastructure.PasswordHasher;
import br.com.portalproposta.backend.model.Perfil;
import br.com.portalproposta.backend.model.Usuario;
import br.com.portalproposta.backend.repository.PerfilRepository;
import br.com.portalproposta.backend.repository.UsuarioRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

@Service
public class UsuarioService {

    private final UsuarioRepository usuarioRepository;
    private final PerfilRepository perfilRepository;

    @Autowired
    public UsuarioService(UsuarioRepository usuarioRepository, PerfilRepository perfilRepository) {
        this.usuarioRepository = usuarioRepository;
        this.perfilRepository = perfilRepository;
    }

    public List<Usuario> list() {
        return usuarioRepository.findAllByOrderByNomeAsc();
    }

    public Usuario getById(String id) {
        return usuarioRepository.findById(id)
                .orElseThrow(() -> new AppNotFoundException("usuário não encontrado"));
    }

    @Transactional
    public Usuario create(Usuario user, String password) {
        if (user.getNome() == null || user.getNome().trim().isEmpty()) {
            throw new AppBadRequestException("nome do usuário é obrigatório");
        }

        if (user.getEmail() == null || user.getEmail().trim().isEmpty() || !user.getEmail().contains("@")) {
            throw new AppBadRequestException("email do usuário é obrigatório ou inválido");
        }

        Perfil perfilObj = perfilRepository.findById(user.getPerfilId())
                .orElseThrow(() -> new AppNotFoundException("perfil de acesso não encontrado"));

        usuarioRepository.findByEmail(user.getEmail()).ifPresent(existing -> {
            throw new AppBadRequestException("já existe um usuário com este email");
        });

        if (password == null || password.trim().isEmpty()) {
            throw new AppBadRequestException("a senha é obrigatória");
        }

        user.setId(UUID.randomUUID().toString());
        user.setSenhaHash(PasswordHasher.hashPassword(password));
        user.setCreatedAt(LocalDateTime.now());
        user.setUpdatedAt(LocalDateTime.now());

        Usuario saved = usuarioRepository.save(user);
        saved.setPerfil(perfilObj);
        return saved;
    }

    @Transactional
    public Usuario update(String id, Usuario input, String newPassword) {
        Usuario user = getById(id);

        if (input.getNome() == null || input.getNome().trim().isEmpty()) {
            throw new AppBadRequestException("nome do usuário é obrigatório");
        }

        if (input.getEmail() == null || input.getEmail().trim().isEmpty() || !input.getEmail().contains("@")) {
            throw new AppBadRequestException("email do usuário é obrigatório ou inválido");
        }

        if (!input.getEmail().equals(user.getEmail())) {
            usuarioRepository.findByEmail(input.getEmail()).ifPresent(existing -> {
                if (!existing.getId().equals(user.getId())) {
                    throw new AppBadRequestException("já existe um usuário com este email");
                }
            });
            user.setEmail(input.getEmail());
        }

        if (input.getPerfilId() != null && !input.getPerfilId().isEmpty()
                && !input.getPerfilId().equals(user.getPerfilId())) {
            Perfil perfilObj = perfilRepository.findById(input.getPerfilId())
                    .orElseThrow(() -> new AppNotFoundException("perfil de acesso não encontrado"));
            user.setPerfilId(input.getPerfilId());
            user.setPerfil(perfilObj);
        }

        user.setNome(input.getNome());
        user.setAtivo(input.isAtivo());

        if (newPassword != null && !newPassword.trim().isEmpty()) {
            user.setSenhaHash(PasswordHasher.hashPassword(newPassword));
        }

        user.setUpdatedAt(LocalDateTime.now());

        Usuario saved = usuarioRepository.save(user);
        if (saved.getPerfil() == null) {
            saved.setPerfil(perfilRepository.findById(saved.getPerfilId()).orElse(null));
        }
        return saved;
    }

    @Transactional
    public void delete(String id) {
        Usuario user = getById(id);

        if (user.getPerfil() != null && "Super Admin".equals(user.getPerfil().getNome())) {
            long count = usuarioRepository.countByPerfilNomeAndAtivo("Super Admin", true);
            if (count <= 1) {
                throw new AppBadRequestException("não é possível excluir o último usuário super admin");
            }
        }

        usuarioRepository.delete(user);
    }

    public Usuario login(String email, String password) {
        Usuario user = usuarioRepository.findByEmail(email)
                .orElseThrow(() -> new AppUnauthorizedException("email ou senha inválidos"));

        if (!user.isAtivo()) {
            throw new AppUnauthorizedException("email ou senha inválidos");
        }

        String hashedInput = PasswordHasher.hashPassword(password);
        if (!user.getSenhaHash().equals(hashedInput)) {
            throw new AppUnauthorizedException("email ou senha inválidos");
        }

        return user;
    }
}

File: backend/Services/UsuarioService.cs
using Microsoft.EntityFrameworkCore;
using PortalProposta.Data;
using PortalProposta.Infrastructure;
using PortalProposta.Models;

namespace PortalProposta.Services;

public class UsuarioService(AppDbContext context)
{
    public async Task<List<Usuario>> ListAsync() =>
        await context.Usuarios.Include(u => u.Perfil).OrderBy(u => u.Nome).ToListAsync();

    public async Task<Usuario> GetByIdAsync(string id) =>
        await context.Usuarios.Include(u => u.Perfil).FirstOrDefaultAsync(u => u.Id == id)
            ?? throw new AppNotFoundException("usuário não encontrado");

    public async Task<Usuario> CreateAsync(Usuario user, string password)
    {
        Validate.Usuario(user);

        if (!await context.Perfis.AnyAsync(p => p.Id == user.PerfilId))
            throw new AppNotFoundException("perfil de acesso não encontrado");

        if (await context.Usuarios.AnyAsync(u => u.Email == user.Email))
            throw new AppBadRequestException("já existe um usuário com este email");

        if (string.IsNullOrWhiteSpace(password))
            throw new AppBadRequestException("a senha é obrigatória");

        user.Id = Guid.NewGuid().ToString();
        user.SenhaHash = PasswordHasher.Hash(password);
        user.CreatedAt = user.UpdatedAt = DateTime.Now;

        context.Usuarios.Add(user);
        await context.SaveChangesAsync();

        user.Perfil = await context.Perfis.FirstOrDefaultAsync(p => p.Id == user.PerfilId);
        return user;
    }

    public async Task<Usuario> UpdateAsync(string id, Usuario input, string? newPassword)
    {
        var user = await context.Usuarios.Include(u => u.Perfil).FirstOrDefaultAsync(u => u.Id == id)
            ?? throw new AppNotFoundException("usuário não encontrado");

        Validate.Usuario(input);

        if (input.Email != user.Email &&
            await context.Usuarios.AnyAsync(u => u.Email == input.Email && u.Id != user.Id))
            throw new AppBadRequestException("já existe um usuário com este email");

        if (!string.IsNullOrEmpty(input.PerfilId) && input.PerfilId != user.PerfilId)
        {
            if (!await context.Perfis.AnyAsync(p => p.Id == input.PerfilId))
                throw new AppNotFoundException("perfil de acesso não encontrado");
            user.PerfilId = input.PerfilId;
        }

        user.Nome = input.Nome;
        user.Email = input.Email;
        user.Ativo = input.Ativo;

        if (!string.IsNullOrWhiteSpace(newPassword))
            user.SenhaHash = PasswordHasher.Hash(newPassword);

        user.UpdatedAt = DateTime.Now;
        await context.SaveChangesAsync();

        user.Perfil = await context.Perfis.FirstOrDefaultAsync(p => p.Id == user.PerfilId);
        return user;
    }

    public async Task DeleteAsync(string id)
    {
        var user = await context.Usuarios.Include(u => u.Perfil).FirstOrDefaultAsync(u => u.Id == id)
            ?? throw new AppNotFoundException("usuário não encontrado");

        if (user.Perfil?.Nome == "Super Admin")
        {
            var count = await context.Usuarios
                .Include(u => u.Perfil)
                .CountAsync(u => u.Perfil!.Nome == "Super Admin" && u.Ativo);

            if (count <= 1)
                throw new AppBadRequestException("não é possível excluir o último usuário super admin");
        }

        context.Usuarios.Remove(user);
        await context.SaveChangesAsync();
    }

    public async Task<Usuario> LoginAsync(string email, string password)
    {
        var user = await context.Usuarios
            .Include(u => u.Perfil)
            .FirstOrDefaultAsync(u => u.Email == email);

        if (user is null || !user.Ativo || user.SenhaHash != PasswordHasher.Hash(password))
            throw new AppUnauthorizedException("email ou senha inválidos");

        return user;
    }
}
